### DISCUSSION TITLE
Sample vscode-php-plugin configuration in LazyVim

### DISCUSSION THREAD

**Discussion by odysseyalive** [LazyVim/LazyVim#645] [Show and tell]
This project came out of nowhere! I'm really impressed with it.

I'm a PHP dev, and I've been having issues with unstable Emacs PHP debug plugins. So I tried vscode for a couple of days. Some of my projects are so big that vscode was moving like a slug in winter.  There was no way I could use it.

So I brushed off my bloated VIM config, looking to optimize it. That's how I found this project; I'm glad I did. By far, this is the best experience of the vscode-php-plugin yet!  You can actually read the variables. 

The LazyVim configuration was a new animal to me. There needs to be better documentation on how the vscode-php-plugin works in Neovim. Although this may be better served for the Mason documentation, I thought I'd also drop it here. This is a LazyVim-centric setup, anyhow. So, here it is:

1. load the php-debug-adapter inside Mason.

2. drop the following code into your LazyVim init.lua:
```
-- the following settins assume you've installed the vscode-debug-php adapter using mason.nvim
local dap = require("dap")
dap.adapters.php = {
  type = "executable",
  command = "node",
  args = {
    vim.loop.os_homedir() .. "/.local/share/nvim/mason/packages/php-debug-adapter/extension/out/phpDebug.js",
  },
}
require("dap.ext.vscode").load_launchjs(vim.fn.getcwd() .. "/.vscode/launch.json", {})
```

3. Next add the launch.json file inside a subdirectory called .vscode in your project root:

```
{
	"version": "0.2.0",
	"configurations": [
		{
			"name": "Use custom launch.json script",
			"type": "php",
			"request": "launch",
			"port": 9003,
			"pathMappings": {
				"/var/www": "/home/username/your-porject-folder
			},
			"sourchMaps": true,
			"log": false
		}
	]
}
```
Your launch script doesn't have to look like this. This is just an example of running the project in a container or VM.

4. Most important of all and not documented anywhere, \<leader\>dc starts the process.

I hope that helps someone. It took me hours to get this working. For whatever reason, the location of the launch.json is being overwritten somewhere along the way. The declaration there in the init file shouldn't be overlooked.

**Comment by jasalt** [comment]
Hello, I found myself in a bit similar situation. Tried to migrate from VSCode to Emacs and couldn't get full PHP indenting working (web-mode / php-mode). Maybe it will get better in a couple years with tree-sitter etc.. Found indenting working better out of the box with nvim so thought to try it out.

Trying this guide as a near total nvim newbie with empty LazyVim starter config.

1. Opening mason with `<leader> cm`, scrolling down to find `php-debug-adapter`, opened view with enter, click it few times figuring how to install, randomly hit `i` on keyboard, fancy notifications shows it installed after a moment.

2. Copied config, relaunching nvim gives error of `require("dap")` failing cause package is missing. I found that I had to first modify `~/.config/nvim/lua/config/lazy.lua` and add line `{ import = "lazyvim.plugins.extras.dap.core" },` according to https://www.lazyvim.org/extras/dap/core and then it worked.

3. Had a pre-existing `.vscode/launch.json` in project root, previously debugging WordPress plugin code with VSCode running inside VVV Vagrant box.

4. Starting process with `<leader>dc` shows it starting and finding the correct `launch.json` having the correct debug profile name. Able to add breakpoints with `<leader>db`. However it does not stop the evaluation on breakpoint as it does on VSCode, so something is still off.

Need to learn more about this later, seems it got pretty far for now. Will continue doing some work learning basic movements with vscode-neovim addon on VSCode side etc..

If you have more info how/if you have got the LazyVim setup working with PHP I would be interested to hear. Thank's for this short guide also.

**Reply by jasalt** [reply]
VSCode felt too laggy and it was weekend anyways so digged into Neovim / LazyVim. Got the PHP LSP and DAP debugger working with [VVV Vagrant](https://varyingvagrantvagrants.org/) setup which maps the whole www/WordPress folder to host. 

There seemed to be some issue with existing `.vscode/launch.json` at www root path being incompatible with the DAP adapter. Inspecting the XDebug logs `/var/log/php/xdebug-remote.log` it was connecting to the port ok, but wasn't resolving the breakpoint locations from `pathMappings` and stopping to set breakpoints. 

```
.vscode/launch.json (not working):
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "VVV Listen for Xdebug",
      "type": "php",
      "request": "launch",
      "port": 9003,
      "pathMappings": {
        "/srv/www/sitename/public_html/": "${workspaceRoot}/"
      }
    }
  ]
}
```

Patched together a working config looking at https://github.com/mfussenegger/nvim-dap/wiki/Debug-Adapter-installation#php and some Reddit threads:

```
-- init.lua (working config)

  local dap = require("dap")
  dap.adapters.php = {
    type = "executable",
    command = "node",
    args = {
      vim.loop.os_homedir() .. "/.local/share/nvim/mason/packages/php-debug-adapter/extension/out/phpDebug.js",
    },
  }
  dap.configurations.php = {
    {
      type = "php",
      request = "launch",
      name = "Listen for Xdebug via DAP",
      port = 9003,
      localSourceRoot = "/home/user/sites/vvv/www/sitename/public_html/",
      serverSourceRoot = "/srv/www/sitename/public_html/",
    },
  }
```

This is working now when starting nvim at www path / WP site root: `nvim /home/user/sites/vvv/www/sitename/public_html/`.

To directly open eg. a plugin folder such as `nvim /home/user/sites/vvv/www/sitename/public_html/wp-content/plugins/my-plugin/` the LSP/DAP `root_dir` still would need to be configured so that it would use the actual WWW/WP root path to find all WP core definitions etc.
`
The config is also site specific, should instead use some project specific launch.json equivalent nvim config file.

**Reply by odysseyalive** [reply]
I posted my setup here: https://github.com/bridgesense/dotfiles/tree/master/workstation/nvim

I use this plugin every day. Still working through one more rough spot with not being able to edit code when dap is terminated without restarting nvim.  I need to file a ticket on it.

I wrote a little documentation around my setup here: https://github.com/bridgesense/dotfiles/tree/master#vim

If you look in the init.lua, I've made some tweaks. My dap configurations are now in each project directory under ".vscode/launch.json"...  _Evidently, that's a requirement._  

So rather than having the dap.configurations.php in the init.lua, I've added this line instead:

```
require("dap.ext.vscode").load_launchjs(vim.fn.getcwd() .. "/.vscode/launch.json", {})
```

Inside each project  is  the .vscode directory with a standard launch.json launch file:

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Listen for Xdebug",
      "type": "php",
      "request": "launch",
      "port": 9003,
      "log": true,
      "pathMappings": {
        "/var/www": "/home/francis/www/x.com"
      }
    }
  ]
}

```

**Reply by jasalt** [reply]
Good to hear you have been having good results with it, I'll definitely keep an eye on these config's, thank's. 

> one more rough spot with not being able to edit code when dap is terminated

Didn't face this yet myself at least when terminating and starting a new debug session and doing some simple changes in between, but I hope it gets solved. 

So far for me the overall feeling with the debugger has been promising for most part, noticed only a couple minor things. 

I had some UI freezes after fiddling with the REPL bar `(<leader> d r)` that was keeping focus after using it and did something mad when receiving bunch of panic keypresses as I had forgotten where the cursor was, it recovered though after 20 sec or so. Secondly the debugger eval function with `<leader> d e` does not seem to work as I expect it would work (evaluates visual region?). Selecting and evaluating `$api_url_demo = "asd";` as example gives "can not get property" error while same expression works ok from REPL or expressions tab.

**Comment by odysseyalive** [comment]
> I had some UI freezes after fiddling with the REPL bar `(<leader> d r)` that was keeping focus after using it and did something mad when receiving bunch of panic keypresses as I had forgotten where the cursor was, it recovered though after 20 sec or so. Secondly the debugger eval function with `<leader> d e` does not seem to work as I expect it would work (evaluates visual region?). Selecting and evaluating `$api_url_demo = "asd";` as example gives "can not get property" error while same expression works ok from REPL or expressions tab.

I've been using it only on a basic level. Although,, this is the first place I've seen the "Run to Cursor" even work.  To start the process, you'll need to run \<leader\> d c, which was completely backwards than any of the similar debugging interfaces I've used before.  I'm still making friends with it myself.  I personally think that \<leader\> d o should be switched around with \<leader\> d O.  I haven't given much thought on how to change it.

**Reply by jasalt** [reply]
I found my way to the DAP UI first (similar to VSCode and Emacs dap-mode) `<leader> d u` and clicked "play" icon with mouse to get debug session running as expected, however I got the tip from your initial post to start it with `<leader> d c` which is convenient. 

I would second that `<leader>do` would make more sense for the more frequently needed step over action. The LazyVim's custom config is shown here under "full spec" https://www.lazyvim.org/extras/dap/core#nvim-dap. Not sure yet how to override these LazyVim extra plugin values in cleanest way though (TODO https://www.lazyvim.org/configuration/plugins#%EF%B8%8F-customizing-plugin-specs). 

```
{ "<leader>do", function() require("dap").step_out() end, desc = "Step Out" },
{ "<leader>dO", function() require("dap").step_over() end, desc = "Step Over" },
```

Missing some "classic" function key debug controls, I added a couple to get started with:

```
vim.api.nvim_set_keymap("n", "<leader><F5>", ":DapContinue<CR>", { noremap = true, silent = true })
vim.api.nvim_set_keymap("n", "<F10>", ":DapStepOver<CR>", { noremap = true, silent = true })
````

I was also missing the mapping for debug UI's restart action (Ctrl + Shift + F5 in VSCode etc.) but it's dubbed "Run last" that has a default LazyVim binding `<leader>dl`.

**Comment by odysseyalive** [comment]
Yeah, whoever put the menu together had never used the plugin. I'm just happy it works, and it isn't broken every other day like it was in Emacs. :)

**Comment by jasalt** [comment]
One thing I was missing was stopping on thrown exceptions which needs to be configured separately as explained in [:help nvim-dap](https://github.com/mfussenegger/nvim-dap/blob/1c63f37f95cd4fb54512898168138d9a75d1516a/doc/dap.txt#L755) "set exception breakpoints". 

Testing this interactively by running `:lua require'dap'.set_exception_breakpoints()` during debug session and accepting the default suggestion it works, debugger is stopping on errors as other editors. Exception info can be read in locals tab while it would be cool to have more prominent notification..

Couldn't figure out yet how to set it on in init.lua config without getting "Cannot set exception breakpoints: No active session!" on nvim startup, so for now just set a couple keybindings to use when debug session is already running, first one skipping deprecation notices:

```
vim.keymap.set("n", "<leader>da", function()
  require("dap").set_exception_breakpoints({ "Warning", "Error", "Exception" })
end, { desc = "Stop on exceptions" }) -- TODO this one doesn't show on which-key

vim.keymap.set("n", "<leader>dA", function()
  require("dap").set_exception_breakpoints({ "Notice", "Warning", "Error", "Exception" })
end, { desc = "Stop on all" })
```

(EDIT: asked in the nvim-dap repo discussions about how to enable on init...)

**Reply by jasalt** [reply]
To stop on exceptions by default on debug session, following can be added to config.

``` 
local dap = require("dap")
dap.defaults.php.exception_breakpoints = { "Notice", "Warning", "Error", "Exception" }
```

Related nvim-dap repo discussion: https://github.com/mfussenegger/nvim-dap/discussions/576#discussioncomment-6782258.

**Comment by EvanDaley** [comment]
Thanks for writing this up. I'd love it if we could get some PHP support into this folder too - https://github.com/LazyVim/LazyVim/tree/main/lua/lazyvim/plugins/extras/lang

Anyone have any ideas for how to get started with that? I could try to do a PR for it but I'm super new to LazyVim and NeoVim in general

**Reply by jasalt** [reply]
Yea, that would make sense. I looked through some of those language configs and would be quite simple to imitate format from there when you get a fully functioning PHP config at hand. I haven't touched my config for a moment but got it into mostly usable state earlier. Still need to polish some rough edges before publishing..

One thing that wasn't mentioned here was configuring Intelephense options for specific code style. I had trouble with function declaration braces auto-indenting to a newline after parentheses eg:

```
function hello_world($arg)
{

}
```

While I wanted:

```
function hello_world($arg){

}
```

This needed some config to be passed to the Intelephense LSP and while VSCode has braces "k&r" option selectable in settings UI (https://github.com/bmewburn/vscode-intelephense/issues/22#issuecomment-688641770), I got this working with NeoVim / LazyVim by adding following to `~/.config/nvim/lua/plugins/lsp.lua`:

```
return {
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      servers = {
        intelephense = {
          settings = {
            intelephense = {
              format = {
                braces = "k&r",
              },
            },
          },
        },
      },
    },
  },
}
```

Still need to look for a way for loose spacing inside parentheses (WordPress style), eg: `( $arg, $arg2 )`, as current auto-indent tightens it as `($arg, $arg2)`...

**Comment by jasalt** [comment]
I uploaded my complete unpolished but working PHP XDebug debugging config here https://github.com/jasalt/nvim. Picture of the interface also here as a proof that it works.

![Screenshot_20231202_202149](https://github.com/LazyVim/LazyVim/assets/2306521/d4f6a24e-fabd-4231-99f8-286cba8ddcdc)
